Feature/operator problem

arbabkhanocloud opened the pull request:
This is JavaScript Task#2 Operator Problem. I converted the prefix expression into an infix expression according to the given pattern. Kindly Check the file (fileName: operator Problem_latest.js)

arbabkhanocloud committed:
Operator Problem

This is javaScript task#2 Operator Problem. Which is converting prefix expression to Infix expression.

arbabkhanocloud committed:
Update OperatorAssignment.js

arbabkhanocloud committed:
Operator problem Task revised

This is the Operator problem task again. In the previous file, I used the operator and operand methods to convert the prefix expression to an infix expression. Now And in this file,
I converted the expression according to the infix pattern.

arbabkhanocloud committed:
Operator problem Task Recursive version

this is latest version and in this file, I converted  the given array into infix expression  by using recursive method. And also changed variables type to  "const" type and made code more elegant and readable as possible.

mahmad35 commented on OperatorAssignment.js:
```diff
@@ -0,0 +1,57 @@
+while( i >= 0 ){
+    if( isOperator(flatedArray[i]) === true ){           // inCase of operator pop two operands and place operator between them and push into stack         
+        let firstOperand = bufferArray[bufferArray.length -1];
+        bufferArray.pop();
```
You can write it as:
- let firstOperand = bufferArray.pop()
- let secondOperand = bufferArray.pop()

mahmad35 commented on OperatorAssignment.js:
```diff
@@ -0,0 +1,57 @@
+        let temp ='';
+        let temp2;
+        if(flatedArray[i] === 'AND' || flatedArray[i] === 'XOR' ){
+            temp2 = temp.concat( '(', firstOperand, ' ', flatedArray[i], ' ', secondOperand, ')' )
```
- const temp = ''.concat('(', firstOperand, ' ', flatedArray[i], ' ', secondOperand, ')');

mahmad35 commented on OperatorAssignment.js:
```diff
@@ -0,0 +1,57 @@
+        case '%':
+            return true;
+    }
+    return false;
```
When you are returning true in the case statement. then you should return false in the switch statement (default).

mahmad35 commented on OperatorAssignment2.js:
```diff
@@ -0,0 +1,48 @@
+let j=1;
+while( i >= 0 ){ 
+    if( bufferArray.length === 3){
+        let firstElement = bufferArray[bufferArray.length -1];
```
Same as in the last file (OperatorAssignment.js)
- let firstElement =  bufferArray.pop()
- let secondElement = bufferArray.pop()

mahmad35 commented on OperatorAssignment2.js:
```diff
@@ -0,0 +1,48 @@
+        let temp ='';
+        let temp2;
+        if(flatedArray[i] ==='AND' ){
+             temp2 = temp.concat('(',firstElement,' ',flatedArray[i],' ',secondElement,')')
```
Same as in the last file (OperatorAssignment.js)
- const tempStr = ''.concat('(',firstElement,' ',flatedArray[i],' ',secondElement,')')
OR
- const tempStr = `(${firstElement} ${flatedArray[i] ${secondElement})}`

mahmad35 commented on OperatorProblem3_latest.js:
```diff
@@ -0,0 +1,62 @@
+  ["AND", ["==", "c", "d"], ["!=", "e", "f"]],
+];
+const newArray = [];
+const infixExpression = recursiveFunction(array); // passing given array to recursiveFunction function
```
There are too many comments here and there.
Only write comments when the function is complex. The comments should tell why this is complex instead of its implementation detail (how it works)

arbabkhanocloud replied:
Noted Ahmed bhai

mahmad35 commented on OperatorProblem3_latest.js:
```diff
@@ -0,0 +1,62 @@
+    for (let i = 0; i < element.length; i++) {
+      if (Array.isArray(element[i])) {
+        newArray.push(recursiveFunction(element[i])); // if next element is array is array then passing it to recursiveFunction function
+        if (newArray.length === 3) {
```
What is the need of this if condition?. It can work without this condition.

arbabkhanocloud replied:
In given array, pattern is that every array have 3 maximum element and these element can also be a nested arrays or  simple element. So I used this " if " check here because whenever all nested elements of an array included (array_element and simple elements) finished then they immediatley pushed into newArray. whenever newArray length equals to 3. then I poped newArray and make infix exprssion and then push it back into newArray which now only contain one infix expression element. and so on ...

arbabkhanocloud replied:
yes I removed this extra line of Code.

mahmad35 commented on OperatorProblem3_latest.js:
```diff
@@ -0,0 +1,62 @@
+          const operand1 = newArray.pop();
+          const operator = newArray.pop();
+          const emptyString = "";
+          const makingInfixExpression = emptyString.concat(
```
const infixExpression = ''.concat(operand1, operator, operand2)

arbabkhanocloud replied:
Noted.

mahmad35 commented on OperatorProblem3_latest.js:
```diff
@@ -0,0 +1,62 @@
+        operator,
+        oprand2
+      );
+      newArray.push(makingInfixExpression);
```
Why are you adding this to the global array and then removing it from the array? you can easily return them with the if-else block.

arbabkhanocloud replied:
I have encapsulated newArray with closure property. So newArray is not global anymore in new file code.

mahmad35 reviewed (changes requested):
I have sone suggestions if you can have a look then i can re-review

arbabkhanocloud committed:
Removed Old files

arbabkhanocloud committed:
updated Operator Problem solution

Encapsulated the newArray with closure property. And removed the extra variable
"emptyString" for concatenation infix Expression.

arbabkhanocloud committed:
Update OperatorProblem_latest.js

Removed Extra line of Code

arbabkhanocloud committed:
Update OperatorProblem_latest.js

changed function names.
